Plugins/NativeEngine/Source/NativeEngine.cpp

```diff
@@ -258,28 +262,68 @@ namespace Babylon
+            };
+
+            auto mem = bgfx::makeRef(image->m_data, image->m_size, releaseFn, image);
+
+            texture->Handle = bgfx::createTexture2D(image->m_width, image->m_height, (image->m_numMips > 1), 1, Cast(image->m_format), BGFX_TEXTURE_NONE | BGFX_SAMPLER_NONE, mem);
+            texture->Width = image->m_width;
+            texture->Height = image->m_height;
+        }
+
+        void CreateCubeTextureFromImages(TextureData* texture, const std::vector<bimg::ImageContainer*>& images, bool hasMips)
         {
-            if (invertY)
+            const bimg::ImageContainer* firstImage = images.front();
+            uint32_t width = firstImage->m_width;
+            uint32_t height = firstImage->m_height;
+            bgfx::TextureFormat::Enum format = Cast(firstImage->m_format);
+
+            uint32_t totalSize = 0;
+            for (auto image : images)
             {
-                FlipYInImageBytes(gsl::make_span(static_cast<uint8_t*>(image->m_data), image->m_size), image->m_height, image->m_size / image->m_height);
+                totalSize += image->m_size;
             }
 
-            if (generateMips)
+            // Combine all the faces into one chunk.
+            const bgfx::Memory* mem = bgfx::alloc(totalSize);
+            uint8_t* ptr = mem->data;
+            for (bimg::ImageContainer* image : images)
             {
-                GenerateMips(allocator, &image);
+                std::memcpy(ptr, image->m_data, image->m_size);
+                ptr += image->m_size;
+                bimg::imageFree(image);
             }
 
-            auto releaseFn = [](void* ptr, void* userData)
-            {
-                bimg::imageFree(static_cast<bimg::ImageContainer*>(userData));
-            };
+            texture->Handle = bgfx::createTextureCube(width, hasMips, 1, format, BGFX_TEXTURE_NONE | BGFX_SAMPLER_NONE, mem);
+            texture->Width = width;
+            texture->Height = height;
+        }
 
-            auto mem = bgfx::makeRef(image->m_data, image->m_size, releaseFn, image);
+        template<typename T>
+        inline napi_ref DetachRef(Napi::Reference<T>& ref)
+        {
+            ref.SuppressDestruct();
+            return static_cast<napi_ref>(ref);
+        }
+
+        inline napi_ref DetachFunctionRef(Napi::FunctionReference& ref)
+        {
+            return DetachRef(ref);
+        }
+
+        template<typename T>
+        inline Napi::Reference<T> AttachRef(napi_env env, napi_ref ref)
+        {
+            return Napi::Reference<T>{env, ref};
+        }
 
-            textureData->Texture = bgfx::createTexture2D(image->m_width, image->m_height, generateMips, 1, Cast(image->m_format), 0, mem);
-            textureData->Width = image->m_width;
-            textureData->Height = image->m_height;
+        inline Napi::FunctionReference AttachFunctionRef(napi_env env, napi_ref ref)
+        {
+            return Napi::FunctionReference{env, ref};
         }
```

Comment:
This stuff seems kind of funky. This is for getting around the non-copyable N-API function thing, right? Can we create a type that encapsulates this behavior and hides it from us?

Reply:
Certainly, but then maybe just fix Napi::Reference to allow copying. I went back and forth on this a bunch. Let me try some things

Reply:
I added a Napi::SharedReference class to do this for now. Maybe we can discuss with NAPI people to see if this works for them. If not, we can still continue to use this in our code.